idseq_dag/steps/blast_contigs.py

```diff
@@ -15,12 +15,103 @@
 MIN_REF_FASTA_SIZE = 25
 MIN_ASSEMBLED_CONTIG_SIZE = 25
 
+# When composing a query cover from non-overlapping fragments, consider fragments
+# that overlap less than this fraction to be disjoint.
+NT_MIN_OVERLAP_FRACTION = 0.1
+
+# NT alginments with shorter length are associated with a high rate of false positives.
+# NR doesn't have this problem because Rapsearch contains an equivalent filter.
+# TODO: Nevertheless, it may be useful to re-filter blastx results.
+NT_MIN_ALIGNMENT_LEN = 36
+
+# Ignore NT local alignments with sequence similarity below 90%.
+NT_MIN_PIDENT = 90
```

Comment:
I agree that this seems to be a reasonable threshold based on experience interpreting this, but out of curiosity - is there any validation on this threshold? I'm curious about examples where this threshold is optimized. 

A similar comment around the NT_MIN_OVERLAP_FRACTION variable.

Reply:
It would be interesting to benchmark the performance based on this NT_MIN_PIDENT; we might need different thresholds for different applications.

I think we should set this parameter to a reasonably high value. The reasoning behind that is: we want to be highly confident to assign an origin of species only for highly similar alignments using blastn, and for distantly related sequences, using blastx/nt would be a better option. Let me know what you think about this. One side note, for longer contigs from metagenomics studies, blastxting the whole contig won't give us any results, when there are two many ORFs on that contigs. One approach is to gene annotation and then blastx. We can tackle this problem later.

Reply:
@katrinakalantar and @zhaoc1    Any clues for samples that would be particularly sensitive to lower or higher values of this threshold?   As initial candidates for validation.

Reply:
I agree that a stringent NT_MIN_PIDENT makes sense in general - and that divergent taxa should rely on the blastx/nr calls! This will especially true if we assign one taxID per read.

The current IDseq workflow may assign a single read to multiple different taxIDs (one in NT and one in NR). It'd be great to think about this threshold wrt the current IDseq workflow - what would happen if a read obtains an initial taxonomic assignment based on GSNAP alignment, with %ident = 80 and then the assembled contig doesn't meet the 90% ident threshold for BLAST...this read would in theory be assigned to the taxid of the initial short read, even if the assembled contig mapped with better %id (id 89%).
Do you have thoughts on this?

Reply:
This is a good question. To answer this question we need to know what percentage of shotgun metagenomics reads can be assigned on a species-level. This paper (https://msystems.asm.org/content/3/6/e00069-18) mentioned only 40%  for stool samples, even lower for some other sample types.  So when we BLAST reads, especially when we only keep the top one hit, the taxonomic classification might not be consistent with the answer when we "assemble" reads into the contigs "context". So even when we have the same sequence similarity threshold, GSNAP alignment for the reads may still be different with the BLAST alignment for the contigs. 

That being said, I think it's might be a good idea to keep GSNAP's %ident same with the BLAST %ident (?), before we do more benchmarks in the future. What do you think?

Reply:
If I understand your concern correctly, @katrinakalantar ,  the NT_MIN_PIDENT threshold for blast results should be *no higher* than the threshold for GSNAP results.   That way, if BLAST finds a better alignment for the contig than GSNAP found for some of the reads, we wouldn't discard the BLAST result and replace it with the inferior GSNAP result.

Reply:
In consideration of this "monotonicity" concern, I am changing the threshold to 80% for the test, to match the GSNAP threshold.   I wonder if we should just have it be derived from the
GSNAP config to prevent mistakes.

Reply:
Thank you @boris-dimitrov - this addresses the concern!